src/universal/components/SignInPage/SignInPage.js

```diff
@@ -0,0 +1,222 @@
+// keys, but works for our own `?redirectTo=/path/to/page` contract.
+const parseSearch = (search: string): Object => {
+  if (!search) { return {}; }
+  const assocArray = chunkArray(search.slice(1).split(/=|&/), 2);
+  return assocArray.reduce((acc, [key, val]) => ({...acc, [key]: val}), {});
+};
+
+// Note that auth0-js seems to break SSR.  Therefore we dynamically import it
+// over the network.
+const LoadableSignInPage = Loadable({
+  loader: () => import('auth0-js'),
+  loading: LoadableLoading,
+  render: (auth0, props) => {
+    const webAuth = new auth0.WebAuth({
+      domain: __AUTH0_DOMAIN__,
+      clientID: __AUTH0_CLIENT_ID__,
+      redirectUri: window.location.href,
+      scope: 'openid rol tms bet'
+    });
+    return <SignInPage webAuth={webAuth} {...props} />;
+  }
+});
+
+class SignInPage extends Component<Props, State> {
+  state = {
+    error: null,
+    loggingIn: false
+  };
+
+  componentWillMount() {
+    this.maybeRedirectToApp();
+    this.maybeCaptureAuthResponse();
+  }
+
+  componentDidUpdate() {
+    this.maybeRedirectToApp();
+  }
+
+  getHandlerForThirdPartyAuth = (auth0Connection: string) => () => {
+    this.props.webAuth.authorize({
+      connection: auth0Connection,
+      responseType: 'token'
+    });
+  };
+
+  maybeRedirectToApp = () => {
+    if (this.props.hasSession) {
+      const parsedSearch = parseSearch(this.props.location.search);
+      const pathToVisit = parsedSearch.redirectTo || '/me';
+      this.props.history.replace(pathToVisit);
+    }
+  };
+
+  maybeCaptureAuthResponse = () => {
+    // If we've received an auth response, log us in
+    const {hash} = this.props.location;
+    if (hash) {
+      this.setState({loggingIn: true});
+      this.parseAuthResponse(hash)
+        .then(this.saveTokens)
+        .catch((error) => {
+          this.setState({error});
+          throw error; // Inform raven
+        });
+    }
+  };
+
+  parseAuthResponse = (hash: string): Promise<ParsedAuthResponse> => {
+    return new Promise((resolve, reject) => {
+      this.props.webAuth.parseHash({hash}, (err, authResult) => {
+        if (err) {
+          return reject(err);
+        }
+        return resolve(authResult);
+      });
+    });
+  };
+
+  resetState = () => {
+    this.setState({loggingIn: false, error: null});
+  };
+
+  saveTokens = (response: ParsedAuthResponse): void => {
+    signinAndUpdateToken(this.props.atmosphere, this.props.dispatch, null, response.idToken);
+  };
+
+  authProviders = [
+    {displayName: 'Google', auth0Connection: 'google-oauth2', iconName: 'google'}
+  ];
+
+  handleSubmitCredentials = ({email, password}: Credentials) => {
+    this.props.webAuth.login({
+      email,
+      password,
+      realm: 'Username-Password-Authentication', // FIXME: extract this as AUTH0_REALM to .env
+      responseType: 'token'
+    }, (error) => {
+      this.setState({error});
+      throw error; // Inform raven
```

Comment:
might be more expressive to call raven.captureException.
Then again, personally i say we don't tell raven. all attempts are logged on auth0, and having a whole bunch of "did i use oauth or my password" failures will just pollute the signal/noise ratio on raven

Reply:
Ok, sure!